File: guests/python/toolbar.py
"""The toolbar scene (tools/scenes/toolbar.steps): the `primary` bit as
window chrome, with no toolbar vocabulary to spell."""

import sys

import kaya

app = kaya.App()

save_enabled = True


def on_toggle_save():
    global save_enabled
    save_enabled = not save_enabled
    can_save.set(save_enabled)


def on_save():
    status.set("saved")


def on_find():
    status.set("found")


def on_export():
    status.set("exported")


with app.window(title="toolbar"):
    status = kaya.signal("ready")
    # Written against the MENU ITEM: the promoted button IS that item.
    can_save = kaya.signal(True)

    # CATALOG PREORDER DECIDES PROMOTION.
    with app.menu("File"):
        # The vocabulary has no save glyph, so `done` is the spelling.
        kaya.item("Save", symbol=kaya.Symbol.DONE, primary=True,
                  enabled=can_save, shortcut="primary+s", on_activate=on_save)
        kaya.item("Export", symbol=kaya.Symbol.FORWARD, on_activate=on_export)

    with app.menu("Edit"):
        kaya.item("Find", symbol=kaya.Symbol.SEARCH, primary=True,
                  on_activate=on_find)
        kaya.item("Replace", symbol=kaya.Symbol.EDIT)

    with app.menu("View"):
        kaya.item("Refresh", symbol=kaya.Symbol.REFRESH)
        kaya.item("Info", symbol=kaya.Symbol.INFO)

    with kaya.column():
        kaya.label(bind=status)  # label#0
        kaya.button("toggle save", on_click=on_toggle_save)  # button#0

sys.exit(app.run())

File: guests/js/toolbar.ts
// The toolbar scene (tools/scenes/toolbar.steps): the `primary` bit as
// window chrome, with no toolbar vocabulary to spell.

import * as kaya from "kaya-gui";

const app = new kaya.App();

let saveEnabled = true;

function onToggleSave(): void {
  saveEnabled = !saveEnabled;
  canSave.set(saveEnabled);
}

function onSave(): void {
  status.set("saved");
}

function onFind(): void {
  status.set("found");
}

function onExport(): void {
  status.set("exported");
}

const { status, canSave } = app.window({ title: "toolbar" }, () => {
  const status = kaya.signal("ready");
  // Written against the MENU ITEM: the promoted button IS that item.
  const canSave = kaya.signal(true);

  // CATALOG PREORDER DECIDES PROMOTION.
  app.menu("File", () => {
    // The vocabulary has no save glyph, so `done` is the spelling.
    kaya.item("Save", { symbol: kaya.Symbol.DONE, primary: true, enabled: canSave, shortcut: "primary+s", onActivate: onSave });
    kaya.item("Export", { symbol: kaya.Symbol.FORWARD, onActivate: onExport });
  });

  app.menu("Edit", () => {
    kaya.item("Find", { symbol: kaya.Symbol.SEARCH, primary: true, onActivate: onFind });
    kaya.item("Replace", { symbol: kaya.Symbol.EDIT });
  });

  app.menu("View", () => {
    kaya.item("Refresh", { symbol: kaya.Symbol.REFRESH });
    kaya.item("Info", { symbol: kaya.Symbol.INFO });
  });

  kaya.column(() => {
    kaya.label({ bind: status }); // label#0
    kaya.button("toggle save", { onClick: onToggleSave }); // button#0
  });
  return { status, canSave };
});

app.run();
